### ISSUE TITLE
Support reduced type inferring from Collection::reduce

### ISSUE THREAD

**Issue by laurentmuller** [weirdan/doctrine-psalm-plugin#154]
Plugin should add capability to infer reduced type from `Collection::reduce` based on assertions in a callback, similarly as Psalm currently does for `array_reduce`. 

This is the the same as #58.

This is the current annotation: https://github.com/doctrine/collections/blob/b1f7c33ebb2d73a4860d6b65a11b9662b9904ebd/src/ReadableCollection.php#L241

For example, the following code must not raise any error:

```php
class Item {
    public float $value = 0.0;
}

/** 
 * @param Collection<array-key, Item> $collection
 */
function foo(Collection $collection): float {
    return $collection->reduce(
        fn (float $carry, Item $item): float => $carry + $item->value,
        0.0
    );
}
```

See example: https://psalm.dev/r/80735e2f8d

The `Collections.phpstub` file must updated to reflect changes.

**Comment by laurentmuller** [comment]
It's correct. But no definition exist for the `reduce()` function in the collections stub.

**Comment by laurentmuller** [comment]
FYI:

Please see how is implemented in  [phpstan/phpstan-doctrine v1.5.5](https://github.com/phpstan/phpstan-doctrine/releases/tag/1.5.5)
